lib/asset/nameconflictpolicy.js

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+
+const PRIVATE = Symbol("PRIVATE");
+
+/**
+ * @typedef {Object} NameConflictPolicyOptions
+ * @property {Boolean} [createVersion=false] Create version on duplicates (truthy value)
+ * @property {String} [versionLabel] Version label to apply to the created/updated file
+ * @property {String} [versionComment] Version comment to apply to the created/updated file
+ * @property {Boolean} [replace=false] True if the existing file should be replaced (truthy value)
+ */
+
+/**
+ * Define how to handle a name conflict
+ */
+class NameConflictPolicy {
+
+    /**
+     * Default name conflict policy, which updates the asset in-place
+     * 
+     * @returns {NameConflictPolicy} Default name conflict policy
+     */
+    static defaultPolicy() {
+        return new NameConflictPolicy();
+    }
+
+    /**
+     * Build a createVersion name conflict policy
+     * 
+     * @param {String} [versionLabel] Version label to apply to the created/updated file
+     * @param {String} [versionComment] Version comment to apply to the created/updated file 
+     * @returns {NameConflictPolicy} Create version name conflict policy
+     */
+    static createVersionPolicy(versionLabel, versionComment) {
+        if (versionLabel && (typeof versionLabel !== "string")) {
+            throw Error(`versionLabel is expected to be a string: ${versionLabel}`);
```

Comment:
I think adding the type to the log would help here, in case version labels are Number with type Number (aka `1`, `2`, ...) instead of numbers as string (aka `'1'`, `'2'`, ...). The difference would be hard to see in logs.

Reply:
I'm going to deal with this in a subsequent PR, since otherwise it will confuse this one.